### PR TITLE
Frontend

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
# React+Redux Blog App Frontend:
Description: Frontend app that consumes the restful Blog API and uses Django Knox Token Authentication.

Installation Instructions:
1. run `npm install`
2. run `npm start`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+
+# Created by https://www.gitignore.io/api/django
+# Edit at https://www.gitignore.io/?templates=django
+
+### Django ###
+*.log
+*.pot
+*.pyc
+__pycache__/
```
Comment (zubair-arbi) [inline .gitignore:11]: this `.gitignore` file should contain artifacts related to react app only. Remove all the irrelevant entries

File: webblog/.babelrc [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+    "presets": ["@babel/preset-env", "@babel/preset-react"],
+    "plugins": ["transform-class-properties"]
+}
```
Comment (zubair-arbi) [inline webblog/.babelrc]: make sure to have a newline at the end of file

File: webblog/frontend/admin.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
```
Comment (zubair-arbi) [inline webblog/frontend/admin.py]: Remove these django related files. Only files related to React should be in this repo.

File: webblog/frontend/src/actions/blogList.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+//DELETE_BLOG
+export const deleteBlog = (id) => (dispatch, getState) => {
+    axios.delete(`/api/blog/${id}/`, tokenConfig(getState))
+        .then(res => {
+            dispatch({
+                type: DELETE_BLOG,
+                payload: id
+            });
+        }).catch(err => console.log(err));
```
Comment (zubair-arbi) [inline webblog/frontend/src/actions/blogList.js]: Show a notification to user, maybe something like `There was an error during logout. Please contact our support team`.

File: webblog/frontend/src/components/accounts/Register.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        username: "",
+        email: "",
+        password: "",
+        password2: ""
+    };
+
+    onSubmit = e => {
+        e.preventDefault();
+        const {username, email, password, password2} = this.state;
+        if (password !== password2) {
```
Comment (zubair-arbi) [inline webblog/frontend/src/components/accounts/Register.js]: Please add proper validations for each field: https://www.telerik.com/blogs/up-and-running-with-react-form-validation

File: webblog/frontend/src/components/accounts/Register.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                <div className="card card-body mt-5">
+                    <h2 className="text-center">Register</h2>
+                    <form onSubmit={this.onSubmit}>
+                        <div className="form-group">
+                            <label>Username</label>
+                            <input
+                                type="text"
+                                className="form-control"
+                                name="username"
+                                onChange={this.onChange}
```
Comment (zubair-arbi) [inline webblog/frontend/src/components/accounts/Register.js]: What purpose this `onChange` event binding is serving here?

File: webblog/frontend/src/components/blog/BlogList.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        const currentBlogList = this.props.blogList.slice(indexOfFirstTodo, indexOfLastTodo);
+
+        const pageNumbers = [];
+        for (let i = 1; i <= Math.ceil(this.props.blogList.length / blogsPerPage); i++) {
+            pageNumbers.push(i);
+        }
+
+        const renderPageNumbers = pageNumbers.map(number => {
+            return (
+                <li style="margin-right: 0.3em; color: blue; user-select: none; cursor: pointer;"
```
Comment (zubair-arbi) [inline webblog/frontend/src/components/blog/BlogList.js]: move styling to a proper file

File: webblog/webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+  module: {
+    rules: [
+      {
+        test: /\.js$/,
+        exclude: /node_modules/,
```
Comment (zubair-arbi) [inline webblog/webpack.config.js]: Add webpack configurations for `webpack.common.config.js`, `webpack.dev.config.js` (with hot reload) and `webpack.prod.config.js` (export project in `build` folder).

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
Add project description along with screenshots of different views.

File: webblog/.babelrc (hunk 2 above)
Comment (Ali-D-Akbar) [reply webblog/.babelrc]: done

File: webblog/frontend/src/actions/blogList.js (hunk 4 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/actions/blogList.js]: I'm currently working on errors and alerts

File: webblog/frontend/src/components/accounts/Register.js (hunk 6 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/accounts/Register.js]: It'll update the state of the respective input field accordingly.

File: webblog/frontend/admin.py (hunk 3 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/admin.py]: these files actually make the frontend. It's a django app that handles all the operations on frontend.

File: webblog/frontend/src/actions/types.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+export const GET_BLOGLIST = "GET_BLOGLIST";
```
Comment (etechdevai) [inline webblog/frontend/src/actions/types.js]: add separates objects for each type of messages... Like one for Blogs related constants and similarly for others

File: webblog/frontend/src/components/accounts/Register.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                                value={password}
+                            />
+                        </div>
+                        <div className="form-group">
+                            <label>Confirm Password</label>
+                            <input
+                                type="password"
+                                className="form-control"
+                                name="password2"
+                                onChange={this.onChange}
```
Comment (etechdevai) [inline webblog/frontend/src/components/accounts/Register.js]: user proper names for functions.

File: webblog/frontend/src/components/blog/BlogList.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    };
+
+    render() {
+        const {currentPage, blogsPerPage} = this.state;
+
+        const indexOfLastTodo = currentPage * blogsPerPage;
+        const indexOfFirstTodo = indexOfLastTodo - blogsPerPage;
+        const currentBlogList = this.props.blogList.slice(indexOfFirstTodo, indexOfLastTodo);
+
+        const pageNumbers = [];
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/BlogList.js]: Use available package/components for pagination

File: webblog/frontend/src/components/blog/BlogList.js (hunk 7 above)
Comment (etechdevai) [inline webblog/frontend/src/components/blog/BlogList.js]: use react syntax to add style attribute to elements

File: webblog/frontend/src/components/blog/Dashboard.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import React, { Fragment } from "react";
+import Form from "./Form";
+import BlogList from "./BlogList";
+
+export default function Dashboard() {
+  return (
+    <Fragment>
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/Dashboard.js]: What is the reason to use Fragment?

File: webblog/frontend/src/components/blog/Form.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+            <div className="card card-body mt-4 mb-4">
+                <h2>Add Blog</h2>
+                <form onSubmit={this.onSubmit}>
+                    <div className="form-group">
+                        <label>Title</label>
+                        <input
+                            className="form-control"
+                            type="text"
+                            name="title"
+                            onChange={this.onChange}
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/Form.js]: use proper naming convention for function

File: webblog/frontend/src/components/blog/LoginForm.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    password: ""
+  };
+
+  static propTypes = {
+    addBlog: PropTypes.func.isRequired
+  };
+
+  onChange = e => this.setState({ [e.target.name]: e.target.value });
+
+  onSubmit = e => {
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/LoginForm.js]: use consistent method definition structure throughout the project

File: webblog/frontend/src/components/layout/Header.js [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+                        onClick={this.props.logout}
+                        className="nav-link btn btn-info btn-sm text-light"
+                    >
+                        Logout
+                    </button>
+                </li>
+            </ul>
+        );
+
+        const guestLinks = (
```
Comment (etechdevai) [inline webblog/frontend/src/components/layout/Header.js]: What's the reason of using guestLinks variable ? did you just copy paste a tutorial to get this task done ?

File: webblog/frontend/src/reducers/auth.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    isAuthenticated: null,
+    isLoading: false,
+    user: null
+};
+
+export default function (state = initialState, action) {
+    switch (action.type) {
+        case USER_LOADING:
+            return {
+                ...state,
```
Comment (etechdevai) [inline webblog/frontend/src/reducers/auth.js]: What is the reason of using ...state syntax ?

File: webblog/frontend/src/reducers/auth.js [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        case LOGIN_SUCCESS:
+        case REGISTER_SUCCESS:
+            localStorage.setItem("token", action.payload.token);
+            return {
+                ...state,
+                ...action.payload,
+                isAuthenticated: true,
+                isLoading: false
+            };
+        case AUTH_ERROR:
```
Comment (etechdevai) [inline webblog/frontend/src/reducers/auth.js]: definition of these use cases missing ?

File: webblog/frontend/src/components/blog/LoginForm.js [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        </form>
+      </div>
+    );
+  }
+}
+
+export default connect(
+  null,
+  { addBlog }
+)(Form);
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/LoginForm.js]: Add space at end of files. This issue is in multiple files.

File: webblog/frontend/src/components/blog/LoginForm.js [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+      <div className="card card-body mt-4 mb-4">
+        <h2>Add Blog</h2>
+        <form onSubmit={this.onSubmit}>
+          <div className="form-group">
+            <label>Title</label>
+            <input
+              className="form-control"
+              type="text"
+              name="title"
+              onChange={this.onChange}
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/LoginForm.js]: Use proper naming convention for functions

File: webblog/frontend/src/components/blog/Form.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
+import React, {Component} from "react";
+import {connect} from "react-redux";
+import PropTypes from "prop-types";
+import {addBlog} from "../../actions/blogList";
+
+export class Form extends Component {
+    static propTypes = {
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/Form.js]: What is the reason of using prop-types ?

File: webblog/frontend/src/components/blog/BlogList.js [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+            currentPage: Number(event.target.id)
+        });
+    };
+
+    render() {
+        const {currentPage, blogsPerPage} = this.state;
+
+        const indexOfLastTodo = currentPage * blogsPerPage;
+        const indexOfFirstTodo = indexOfLastTodo - blogsPerPage;
+        const currentBlogList = this.props.blogList.slice(indexOfFirstTodo, indexOfLastTodo);
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/BlogList.js]: is there any other method to find current Blog list ?

File: webblog/frontend/src/components/blog/BlogList.js [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+
+    static propTypes = {
+        blogList: PropTypes.array.isRequired,
+        getBlogList: PropTypes.func.isRequired,
+        deleteBlog: PropTypes.func.isRequired
+    };
+
+    state = {
+        currentPage: 1,
+        blogsPerPage: 3
```
Comment (etechdevai) [inline webblog/frontend/src/components/blog/BlogList.js]: move these constants to environment variables file or constants files

**Review by etechdevai — CHANGES_REQUESTED** [review]

**Comment by etechdevai** [comment]
> _No description provided._

Add proper PR description and PR title

File: webblog/frontend/src/components/blog/BlogList.js (hunk 7 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/blog/BlogList.js]: done

File: webblog/frontend/src/components/accounts/Register.js (hunk 10 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/accounts/Register.js]: this function literally handles the changes occurring in the tag.

File: webblog/frontend/src/components/blog/BlogList.js (hunk 7 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/blog/BlogList.js]: done

File: webblog/frontend/src/components/blog/Dashboard.js (hunk 12 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/blog/Dashboard.js]: Fragment is used to combine multiple children into a single tag without adding every single component to be added separately to the DOM.

ref: https://reactjs.org/docs/fragments.html

File: webblog/frontend/src/components/layout/Header.js (hunk 15 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/layout/Header.js]: I took instructions from the internet. I just separated this code into two parts to reduce redundancy.

File: webblog/frontend/src/reducers/auth.js (hunk 16 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/reducers/auth.js]: this reducer set rest of the states to redux state store.

File: webblog/frontend/src/reducers/auth.js (hunk 17 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/reducers/auth.js]: It had been added.

File: webblog/frontend/src/components/blog/Form.js (hunk 20 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/blog/Form.js]: it is used to require a function addBlog or else it'll give an error before loading the component, which is better than not loading the component.

File: webblog/frontend/src/components/blog/BlogList.js (hunk 21 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/blog/BlogList.js]: this mechanism is used to show only that part of the blog list which is in the page.

File: webblog/frontend/src/components/blog/BlogList.js (hunk 22 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/blog/BlogList.js]: these are the initial states, they're updated for pagination.

File: webblog/frontend/src/components/accounts/Register.js (hunk 5 above)
Comment (Ali-D-Akbar) [reply webblog/frontend/src/components/accounts/Register.js]: Done. The fields are now properly validated.

**Pull request merged**